L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  // registered for this file descriptor, and returns the file descriptor.
+  virtual int ReleaseFD() = 0;
+
+  // Sets various priorities for this file descriptor. The priorities are ignored 
+  // if the Descriptor does not support Priority.
+  virtual void SetReadPriority(Priority priority) = 0;
+  virtual void SetWritePriority(Priority priority) = 0;
+  virtual void SetErrorPriority(Priority priority) = 0;
+
+  // Registers a closure to be run immediately in a worker thread of the 
+  // EventManager, when the file descriptor becomes readable/writable/has error.
+  virtual void NotifyWhenReadable(std::function<void()> upcall) = 0;
+  virtual void NotifyWhenWritable(std::function<void()> upcall) = 0;
+  virtual void NotifyWhenHasError(std::function<void()> upcall) = 0;
+
+  enum NotificationOutcome {
+    NOTIFY_FD_BECAME_READY,
+    NOTIFY_TIMED_OUT
+  };
+
+  // Similar to NotifyWhen*, but with a deadline. If the file descriptor state 
+  // changes before the deadline, then upcall will be called with 
+  // NOTIFY_FD_BECAME_READY. If the deadline expires before the descriptor state 
+  // changes, then upcall will be called with NOTIFY_TIMED_OUT.
+  virtual void NotifyWhenReadableOrTimeout(std::function<void(NotificationOutcome)> upcall, gpr_timespec deadline) = 0;
+  virtual void NotifyWhenWritableOrTimeout(std::function<void(NotificationOutcome)> upcall, gpr_timespec deadline) = 0;
+
+  // Clears the descriptor's status. This allows the NotifyWhen* calls to 
+  // distinguish between old and new events.
+  virtual void ClearReadable() = 0;
+  virtual void ClearWritable() = 0;
+  virtual void ClearHasError() = 0;
+
+  // Forces the descriptor’s status to change, and runs the closures if registered. 
+  // This is especially useful for tearing down closures associated with this 
+  // descriptor by forcing them to run without waiting for the event on the 
+  // underlying socket.
+  virtual void ForceReadable() = 0;
+  virtual void ForceWritable() = 0;
+  virtual void ForceHasError() = 0;
+
+  // Returns a human-readable description of the file descriptor state.
+  virtual string ToString() const = 0;
+};
+
+class EpollEventManagerInterface: public BaseEventManagerInterface {
+ public:
+  virtual ~EpollEventManagerInterface();
+
+  virtual EpollDescriptorInterface* RegisterFileDescriptor(int fd) = 0;
+
+  virtual void DeleteDescriptor(EpollDescriptorInterface* d) = 0;
+
+  // Adds a new one-shot alarm with a slack, which will be scheduled at the given 
+  // deadline plus the timespan specified in slack. Reducing the precision of 
+  // alarms through a timer slack allows the EventManager to perform 
+  // timer-coalescing, which minimizes thread wake-up and has the potential of 
+  // substantially reducing CPU utilization.
+  virtual void ScheduleAlarmInWithSlack(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, TaskKey* key, gpr_timespec slack) = 0;
```

Comment:
`ScheduleAlarmInWithSlack` doesn't look like it's specific to epoll. Many other platforms support timer coalescing.

Should it belong in the base interface? Should it belong in its own interface that can be queried for from the base interface?

Reply:
It's fine to be in the base interface. Thanks for the suggestion.